mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp

```diff
@@ -1560,11 +1574,32 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, tensor::UnPackOp unpackOp,
 
   ArrayRef<int64_t> innerDimPos = unpackOp.getInnerDimsPos();
   ArrayRef<int64_t> innerTiles = unpackOp.getStaticInnerTiles();
+  ArrayRef<int64_t> sourceShape = unpackTensorType.getShape();
+  bool useInBoundsInsteadOfMasking = false;
+  ArrayRef<int64_t> outerDimsPerm = unpackOp.getOuterDimsPerm();
+
+  auto destSize = unpackOp.getDestRank();
+
+  // initVectorShape is the shape of the vector that will be read from the
+  // source tensor. It is set like this: Let's say the sourceShape is 'M' and
+  // the vectorSize (VS) array is size 'N' where N <= M. Thus:
+  // - initVectorShape = sourceShape.take_front(N)
+  // - if outer_dims_perms is present: do that permutation on initVectorShape.
+  // - Multiply all the locations pointed by innerDimPos by the innerTileSize
+  //  attribute value.
+  SmallVector<int64_t> initVectorShape(sourceShape.take_front(destSize));
+  if (inputVectorSizes.empty()) {
+    if (!outerDimsPerm.empty())
+      applyPermutationToVector(initVectorShape, outerDimsPerm);
+    for (auto [i, pos] : llvm::enumerate(innerDimPos))
+      initVectorShape[pos] *= innerTiles[i];
+
+    inputVectorSizes = initVectorShape;
+    useInBoundsInsteadOfMasking = true;
+  }
 
   SmallVector<int64_t> readMaskShape(inputVectorSizes.begin(),
                                      inputVectorSizes.end());
```

Comment:
At this point `useInBoundsInsteadOfMasking` is already set - why do we bother defining and calculating `readMaskShape` if `useInBoundsInsteadOfMasking` is true? It feels like something that should be factored out to a dedicated hook, e.g. `computeReadMaskShapeForUnpackOp`, and then:
```cpp
SmallVector<int64_t> readMaskShape;
if (!useInBoundsInsteadOfMasking)
  readMaskShape = computeReadMaskShapeForUnpackOp();  
```

Also, it looks like the shape calculation for `readMaskShape` and `initVectorShape` are duplicated? Again, why not introduce a dedicated hook for that?

Reply:
Well, I think `readMaskShape` is not named properly. It's actually `readVectorSizes` and we have to unconditionally define it.

Reply:
Yes, we need to calculate it once. I've renamed this to `readVectorSizes` since the mask is generated by createReadOrMaskedRead based on the vectorSizes. The same goes for `writeMaskShape`.